### ISSUE TITLE
Regression in 0.0.78: valid override rejected after a protocol-bound identity decorator

### ISSUE THREAD

**Issue by jelle-openai** [astral-sh/ty#4459]
### Summary

The following self-contained example passes with ty 0.0.75 and 0.0.77, but ty 0.0.78 reports `invalid-method-override` for `Hutia.gnaw`. The decorator returns its input unchanged, and both methods accept the same arguments and return `None`.

```python
from typing import Callable, Protocol

class Capybara(Protocol):

    def gnaw(self):
        pass

def chinchilla[C: Capybara](func: Callable[[C], None]) -> Callable[[C], None]:
    return func

class Degu:

    @chinchilla
    def gnaw(self):
        pass

class Hutia(Degu):

    def gnaw(self):
        pass
```

Run with `ty check --python-version 3.12 repro.py`.

Expected: no diagnostics.

Actual with ty 0.0.78 (`dca9f9873`):

```text
error[invalid-method-override]: Invalid override of method `gnaw`: Definition is incompatible with `Degu.gnaw`
```

Removing `@chinchilla`, or changing the bound from `C: Capybara` to `C: object`, makes both versions pass. No third-party dependencies are required. This was reduced from an otherwise clean project check; every retained reduction was checked against both versions, followed by isolated configuration and import audits.

Environment: macOS arm64; target Python 3.12. The original diagnostic also occurred in Linux CI.

**Comment by carljm** [comment]
This regression is caused by https://github.com/astral-sh/ruff/pull/28156. The cycle here is that in order to test whether `Degu` is assignable to the `Capybara` upper bound of the `C` typevar of `chinchilla`, we have to resolve the effect of `chinchilla` on `Degu.gnaw`. With https://github.com/astral-sh/ruff/pull/28156 that now leaves an unsimplified intersection in the resolved signature `(Degu & Top[Capybara], /) -> None`, which then seems to trigger a signature assignability failure.
